### ISSUE TITLE
Add a way to check for the presence of a Python module

### ISSUE THREAD

**Issue by ebassi** [mesonbuild/meson#3551]
Some projects depend on the availability of a Python module for things like generating documentation or running tests. Typically, with Autotools, you'd use `AX_PYTHON_MODULE`. It would be useful to have something like this for Meson as well.

I'm not entirely sure about the shape this would take; either a method on the `python|python3` modules, e.g.

```meson
  python3 = import('python3')
  python3.find_module('dbusmock', version: '>= 0.10')

  python2 = import('python2')
  mako = python2.find_module('mako', required: false)
  build_docs = mako.found()
```

Or have a dependency object:

```meson
dbusmock = dependency('dbusmock', version: '>= 0.10', modules: 'python3')
mako = dependency('mako', modules: 'python', required: false)
```

**Comment by nirbheek** [comment]
CCing @MathieuDuponchelle

**Comment by nirbheek** [comment]
Personally, I like the `has_module()` approach, which would return a `bool`. Not sure if returning a dependency object is useful or meaningful.

**Comment by ebassi** [comment]
@nirbheek Yeah, a dependency object is kind of wasted on this; it's not like it's going to be used like any other dependency objects. Unless, of course, we think we should rebuild the documentation if the installed version of mako changes, which would be slightly weird but not completely unexpected.

**Comment by MathieuDuponchelle** [comment]
I think going through dependency() would make more sense, as far as I understand that is the purpose of the module argument, and while the dependency object would obviously not bring in any compile-time arguments, it could indeed be passed through depends to build targets that need to be rebuilt when the module changes or is removed.

**Comment by nbenitez** [comment]
Adding here two ways to detect python modules, found on `meson.build` files of current projects. So we can choose the preferred method to implement this feature:

https://gitlab.gnome.org/World/PasswordSafe/blob/master/meson.build#L30
https://gitlab.gnome.org/GNOME/gtk-doc/blob/master/meson.build#L40

**Comment by MathieuDuponchelle** [comment]
@nbenitez , note that the problem here isn't the implementation of the feature, but rather the exposed interface :)

**Comment by a-andre** [comment]
The check if a python module is available was [added to version 0.51.0](https://github.com/mesonbuild/meson/blob/master/docs/markdown/Release-notes-for-0.51.0.md#new-modules-kwarg-for-pythonfind_installation). The only thing missing is a way to specify the required version of the dependency.

**Comment by MathieuDuponchelle** [comment]
@a-andre : sweet, what was the MR ooc?

**Comment by dcbaker** [comment]
I added it, but I don't remember the number of anything.

I had a version that had a `find_module` (or something method) but there were some concerns about it (I don't remember what now).

**Comment by a-andre** [comment]
>  what was the MR ooc?

https://github.com/mesonbuild/meson/pull/5204

**Comment by MathieuDuponchelle** [comment]
thanks! re version checking, I suppose we could extend the current API to accept either strings or dicts?

**Comment by eli-schwartz** [comment]
Checking for a python module cannot guarantee how to find a version for it. Unlike importlib.metadata Distribution objects, they might not have a version at all. If they do, it's only by convention that `__version__` will be the relevant attribute.

**Comment by dcbaker** [comment]
My original proposal had something like (pseudo code because I don't remember exactly):
```python
def find_module(mod: str, version_attr: str = '__version__') -> T.Tuple[bool, str]: ...
```
But there was a lot of concern about what that should look like so I took the path of least resistance.

**Comment by naveen521kk** [comment]
How about using [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html) for py3.8+ or a fallback [`importlib_metadata`](https://importlib-metadata.readthedocs.io) for `<=3.8`? The only thing to consider is that for `py3.8` https://pypi.org/project/importlib-metadata/ must be installed beforehand for this to work.

If that looks good, I can work on implementing that feature.

**Comment by eli-schwartz** [comment]
Meson cannot depend on any modules installed beforehand, even "just on old versions of python".

There are other improvements blocked on importlib.* modules not being available "yet"...

**Comment by eli-schwartz** [comment]
Also we ***cannot*** add version metadata checks to the current implementation (the "modules" kwarg), because it accepts `import` names, not `pip install` names and those aren't always the same. Any suggestions to use importlib.metadata to find versions, must also use it to find "packages", perhaps with a "packages" kwarg...

**Comment by naveen521kk** [comment]
> Meson cannot depend on any modules installed beforehand, even "just on old versions of python".

Hmm, that's not a hard dependency though, it is only required if you want to find a python package and using `py<3.8`.

> There are other improvements blocked on importlib.* modules not being available "yet"...

Wdym?

**Comment by naveen521kk** [comment]
What I think of is having a script like
```py
try:
    # only 3.8+
    from importlib.metadata import version
except (ImportError,ModuleNotFoundError):
    # try using backport
    from importlib_metadata import version
    # fail if it is not available
print(version('{package_name}'))
```
and run it. If it errors out the package isn't found, if it didn't error the package is found the version is printed which will then be compared. If it is `py3.7` and the dependency(`importlib_metadata`) isn't found, then it is it simply error's out saying that the package isn't found.

> Any suggestions to use importlib.metadata to find versions, must also use it to find "packages", perhaps with a "packages" kwarg...

Rather than kwarg I like having a `.find_module()` method as in the original comment(https://github.com/mesonbuild/meson/issues/3551#issue-321515634).

**Comment by eli-schwartz** [comment]
> Hmm, that's not a hard dependency though, it is only required if you want to find a python package and using `py<3.8`.

meson development policy doesn't say anything about "hard dependencies" vs. "only required if you want to use X feature".

**Comment by naveen521kk** [comment]
> meson development policy doesn't say anything about "hard dependencies" vs. "only required if you want to use X feature".

Hmm, then this feature would never be possible. Even when meson is 3.8+ it would support py2...

**Comment by eli-schwartz** [comment]
meson doesn't run on python2, `pymod.find_installation()` won't find it.

It can be documented to not work with `pymod.find_installation('python2')`.

**Comment by naveen521kk** [comment]
> meson doesn't run on python2, pymod.find_installation() won't find it.

Ok

**Comment by lazka** [comment]
What about something like `import('qt5').has_tools()` ->`installation.supports_packages()` ?

**Comment by naveen521kk** [comment]
> What about something like `import('qt5').has_tools()` ->`installation.supports_packages()` ?

LGTM, I can implement that.
